Firestore/core/src/firebase/firestore/util/error_apple.h

```diff
@@ -38,48 +43,45 @@ namespace firestore {
 namespace util {
 
 // Translates a set of error_code and error_msg to an NSError.
-inline NSError* MakeNSError(const int64_t error_code,
-                            const absl::string_view error_msg,
-                            NSError* cause = nil) {
-  if (error_code == FirestoreErrorCode::Ok) {
-    return nil;
-  }
-
-  NSMutableDictionary<NSString*, id>* user_info =
-      [NSMutableDictionary dictionary];
-  user_info[NSLocalizedDescriptionKey] = WrapNSString(error_msg);
-  if (cause) {
-    user_info[NSUnderlyingErrorKey] = cause;
-  }
-
-  return [NSError errorWithDomain:FIRFirestoreErrorDomain
-                             code:static_cast<NSInteger>(error_code)
-                         userInfo:user_info];
-}
+NSError* MakeNSError(const int64_t error_code,
+                     const absl::string_view error_msg,
+                     NSError* cause = nil);
 
 inline NSError* MakeNSError(const util::Status& status) {
   return status.ToNSError();
 }
 
-inline Status MakeStatus(NSError* error) {
-  if (!error) {
-    return Status::OK();
-  }
+Status MakeStatus(NSError* error);
+
+using VoidErrorBlock = void (^)(NSError* _Nullable error);
+
+util::StatusCallback MakeCallback(VoidErrorBlock _Nullable block);
 
-  HARD_ASSERT(error.domain == FIRFirestoreErrorDomain,
-              "Can only translate a Firestore error to a status");
-  auto error_code = static_cast<int>(error.code);
-  HARD_ASSERT(error_code >= FirestoreErrorCode::Cancelled &&
-                  error_code <= FirestoreErrorCode::Unauthenticated,
-              "Unknown error code");
-  return Status{static_cast<FirestoreErrorCode>(error_code),
-                MakeString(error.localizedDescription)};
+template <typename T>
+using VoidValueErrorBlock = void (^)(T _Nullable value,
+                                     NSError* _Nullable error);
+
+template <typename T>
+util::StatusOrCallback<T> MakeCallback(VoidValueErrorBlock<T> _Nullable block) {
+  if (block) {
+    return [block](StatusOr<T> maybe_value) {
+      if (maybe_value.ok()) {
+        block(std::move(maybe_value).ValueOrDie(), nil);
+      } else {
+        block(nil, MakeNSError(std::move(maybe_value).status()));
+      }
+    };
+  } else {
```

Comment:
Question: is this just a fail-safe, or do we have known cases where this is required?